[quant] Quantized adaptive_avg_pool3d

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40271 [quant] Quantized adaptive_avg_pool3d**

Closes #40244

Differential Revision: [D22134318](https://our.internmc.facebook.com/intern/diff/D22134318)

Commit:
[quant] Quantized adaptive_avg_pool3d

[ghstack-poisoned]

Commit:
Update on "[quant] Quantized adaptive_avg_pool3d"


Closes #40244

[ghstack-poisoned]

Commit:
Update on "[quant] Quantized adaptive_avg_pool3d"


Closes #40244

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -1305,6 +1315,115 @@ def test_adaptive_avg_pool2d_nhwc(self, X, output_size_h, output_size_w):
+           output_size_d=st.integers(1, 10),
+           output_size_h=st.integers(1, 10),
+           output_size_w=st.integers(1, 10))
+    def test_adaptive_avg_pool3d(self, X, output_size_d, output_size_h,
```
there are many code duplications in these test, would be great if we can dedup some, I think dedup this one with ndhwc would be striaghtforward? also dedup with the 2d test as well if possible(maybe part of the code).

z-a-f replied:
I partially refactored to remove some dups, but the ndhwc is much harder -- will refactor later

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp:
```diff
@@ -296,5 +448,29 @@ Tensor quantized_adaptive_avg_pool2d(
   return output;
 }
 
+Tensor& quantized_adaptive_avg_pool3d_out(
```
nit: typical argument order should probably be input, output? why does output come before input in this case?

jerryzh168 replied:
I think the order should probably be: input, output_size, output

z-a-f replied:
That what I did initially, but then I saw that out variants for `native_functions` always put output first, so that I did

z-a-f replied:
Also, style guides agree that it should be input->output. I will see if it is possible :)

z-a-f replied:
I checked -- in this particular case (out variant) the output has to be first. For all other cases I refactored.

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5767,13 +5767,15 @@
   dispatch:
     CPU: adaptive_avg_pool3d_cpu
     CUDA: adaptive_avg_pool3d_cuda
+    QuantizedCPU: quantized_adaptive_avg_pool3d
```
I feel we should actually rename this, it should be `adaptive_avg_pool3d_quantized_cpu` same for others

z-a-f replied:
I think I can do it in a separate PR -- it involves changing other ones. WIll creaete a task for myself

jerryzh168 replied:
yeah we don't need to it in this PR, sounds good, thanks

z-a-f replied:
#40315

z-a-f replied:
#41071

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -1337,6 +1337,104 @@ void qadaptive_avg_pool2d_nhwc_kernel(
+          auto* internal_i_p = i_p + istartH * istrideH + istartW * istrideW;
+
+          // TODO: more vectorization with loop interleaving
+          do_avg_pool_on_AVX2<scalar_t>(
```
not related to this diff, but it's non intuitive that this function is a no-op if AVX2 is not available.  Maybe we can add a brief comment here to explain?

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -1337,6 +1337,104 @@ void qadaptive_avg_pool2d_nhwc_kernel(
+  });
+}
+
 void qavg_pool2d_nhwc_kernel(
```
I didn't read this code line by line, but I'm guessing it's not easily reusable/extensable for the 3d case?

z-a-f replied:
I tried reusing it, but it gets really convoluted. I will make a low-pri task to reuse the code, which will require refactoring :(

z-a-f replied:
#40316

z-a-f replied:
I was looking into this a little more -- I could reuse it by only writing the 3d version of it, and when using the 2d version, just set the extra dimension to 1. The problem is that this introduces extra computation steps (like adding zeros or multiplying by 1). Do we need that?

jerryzh168 replied:
I think we can worry about code reuse later, we want to make it work first

vkuzo replied:
makes sense, thanks for the context!

optional suggestion, if the new func is a copy and modify of the existing `qavg_pool2d_nhwc_kernel` function, can we keep the relevant variable names consistent?  would make it easier to realize that it's a copy and to refactor later

jerryzh168 commented on torch/nn/quantized/functional.py:
```diff
@@ -94,6 +94,24 @@ def adaptive_avg_pool2d(input, output_size):
+                     double-integer tuple)
+    """
+    if not input.is_quantized:
+        raise ValueError("Input to 'quantized.adaptive_avg_pool3d' must be quantized!")
```
nit: `functional.quantized. adaptive_avg_pool3d`

jerryzh168 reviewed (approved):
Please fix CI errors before land

Commit:
Update on "[quant] Quantized adaptive_avg_pool3d"


Closes #40244

Differential Revision: [D22134318](https://our.internmc.facebook.com/intern/diff/D22134318)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantized adaptive_avg_pool3d"


Closes #40244

Differential Revision: [D22134318](https://our.internmc.facebook.com/intern/diff/D22134318)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantized adaptive_avg_pool3d"


Closes #40244

Differential Revision: [D22134318](https://our.internmc.facebook.com/intern/diff/D22134318)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantized adaptive_avg_pool3d"


Closes #40244

Differential Revision: [D22134318](https://our.internmc.facebook.com/intern/diff/D22134318)

[ghstack-poisoned]
